fastddsspy_tool/src/cpp/tool/Backend.cpp

```diff
@@ -0,0 +1,96 @@
+// Copyright 2023 Proyectos y Sistemas de Mantenimiento SL (eProsima).
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "Backend.hpp"
+
+namespace eprosima {
+namespace spy {
+
+Backend::Backend(const yaml::Configuration& configuration)
+    : configuration_(configuration)
+    , payload_pool_(std::make_shared<ddspipe::core::FastPayloadPool>())
+    , discovery_database_(std::make_shared<ddspipe::core::DiscoveryDatabase>())
+    , allowed_topics_(
+        std::make_shared<ddspipe::core::AllowedTopicList>(
+            configuration.allowlist,
+            configuration.blocklist))
+    , thread_pool_(
+        std::make_shared<utils::SlotThreadPool>(
+            configuration.n_threads))
+    , participant_database_(std::make_shared<ddspipe::core::ParticipantsDatabase>())
+    , model_(
+        std::make_shared<participants::SpyModel>())
+    , dds_participant_(
+        std::make_shared<participants::SpyDdsParticipant>(
+            configuration.simple_configuration,
+            payload_pool_,
+            discovery_database_))
+    , spy_participant_(
+        std::make_shared<participants::SpyParticipant>(
+            configuration.spy_configuration,
+            payload_pool_,
+            discovery_database_,
+            model_))
+{
+    // Check configuration
+    utils::Formatter error_msg;
+    if (!configuration_.is_valid(error_msg))
+    {
+        throw utils::ConfigurationException(error_msg);
+    }
+
+    // Init Dds Participant
+    dds_participant_->init();
```

Comment:
Why do you need to initialise the participant?

Reply:
Good question. It is due to TSAN and listeners. In short, if an object works as its own listener, it must set the listener before finishing the constructor, and this is a TSAN issue as you can use such object before it has been completely constructed. After many iterations over this problem, we decided that the best solution was to separate creation and initialization in two steps (which goes against RAII, but it was the only solution).